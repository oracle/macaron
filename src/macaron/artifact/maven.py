# Copyright (c) 2024 - 2024, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""This module declares types and utilities for Maven artifacts."""
import re
from collections.abc import Sequence

from packageurl import PackageURL

from macaron.config.defaults import defaults
from macaron.slsa_analyzer.provenance.intoto import InTotoPayload
from macaron.slsa_analyzer.provenance.intoto.v01 import InTotoV01Subject
from macaron.slsa_analyzer.provenance.intoto.v1 import InTotoV1ResourceDescriptor
from macaron.slsa_analyzer.provenance.slsa import extract_build_artifacts_from_slsa_subjects, is_slsa_provenance_payload
from macaron.slsa_analyzer.provenance.witness import (
    extract_build_artifacts_from_witness_subjects,
    is_witness_provenance_payload,
    load_witness_verifier_config,
)


class MavenSubjectPURLMatcher:
    """A matcher matching a PURL identifying a Maven artifact to a provenance subject."""

    @staticmethod
    def get_subject_in_provenance_matching_purl(
        provenance_payload: InTotoPayload, purl: PackageURL
    ) -> InTotoV01Subject | InTotoV1ResourceDescriptor | None:
        """Get the subject in the provenance matching the PURL.

        In this case where the provenance is assumed to be built from a Java project,
        the subject must be a Maven artifact.

        Parameters
        ----------
        provenance_payload : InTotoPayload
            The provenance payload.
        purl : PackageURL
            The PackageURL identifying the matching subject.

        Returns
        -------
        InTotoV01Subject | InTotoV1ResourceDescriptor | None
            The subject in the provenance matching the given PURL.
        """
        if not purl.namespace:
            return None
        if not purl.version:
            return None
        if purl.type != "maven":
            return None

        artifact_subjects: Sequence[InTotoV01Subject | InTotoV1ResourceDescriptor] = []
        if is_witness_provenance_payload(
            payload=provenance_payload,
            predicate_types=load_witness_verifier_config().predicate_types,
        ):
            artifact_subjects = extract_build_artifacts_from_witness_subjects(provenance_payload)
        elif is_slsa_provenance_payload(
            payload=provenance_payload,
            predicate_types=defaults.get_list(
                "slsa.verifier",
                "predicate_types",
                fallback=[],
            ),
        ):
            artifact_subjects = extract_build_artifacts_from_slsa_subjects(provenance_payload)
        else:
            return None

        for subject in artifact_subjects:
            subject_name = subject["name"]
            if not subject_name:
                continue
            _, _, artifact_filename = subject_name.rpartition("/")
            subject_purl = create_maven_purl_from_artifact_filename(
                artifact_filename=artifact_filename,
                group_id=purl.namespace,
                version=purl.version,
            )
            if subject_purl == purl:
                return subject

        return None


def create_maven_purl_from_artifact_filename(
    artifact_filename: str,
    group_id: str,
    version: str,
) -> PackageURL | None:
    """Create a Maven PackageURL given an artifact filename, a group id, and a version.

    For reference, see:
    - https://maven.apache.org/ref/3.9.6/maven-core/artifact-handlers.html
    - https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#maven
    Notes:
    - For the time being, we are only supporting the ``"type"`` qualifier, although the
    Maven section in the PackageURL docs also mention the ``"classifier"`` qualifier.
    This is because not all artifact types has a unique value of ``"classifier"``
    according to the Artifact Handlers table in the Maven Core reference. In addition,
    not supporting the ``"classifier"`` qualifier at the moment simplifies the
    implementation for PURL decoding and generation until there is a concrete use
    case for this additional qualifier.
    - We are only supporting only 4 artifact types: jar, pom, javadoc, and java-source.

    Parameters
    ----------
    artifact_filename : str
        The filename of the artifact.
    group_id : str
        The group id of the artifact.
    version : str
        The version of the artifact.

    Returns
    -------
    PackageURL | None
        A Maven artifact PackageURL, or `None` if the filename does not follow any
        of the supported artifact name patters.
    """
    # Each artifact name should follow the pattern "<artifact-id>-<suffix>"
    # where "<suffix>" is one of the following.
    suffix_to_purl_qualifiers = {
        f"-{version}.jar": {"type": "jar"},
        f"-{version}.pom": {"type": "pom"},
        f"-{version}-javadoc.jar": {"type": "javadoc"},
        f"-{version}-sources.jar": {"type": "java-source"},
    }

    for suffix, purl_qualifiers in suffix_to_purl_qualifiers.items():
        if artifact_filename.endswith(suffix):
            artifact_id = artifact_filename[: -len(suffix)]
            return PackageURL(
                type="maven",
                namespace=group_id,
                name=artifact_id,
                version=version,
                qualifiers=purl_qualifiers,
            )

    return None


def is_valid_maven_group_id(group_id: str) -> bool:
    """Check if the provided string is a valid maven group id.

    Parameters
    ----------
    group_id : str
        The group id to check.

    Returns
    -------
    bool
        True if the group id is valid, False otherwise
    """
    # Should match strings like org.example.foo, org.example-2.foo.bar_1.
    pattern = r"^[a-zA-Z][a-zA-Z0-9-]*\.([a-zA-Z][a-zA-Z0-9-]*\.)*[a-zA-Z][a-zA-Z0-9-]*[a-zA-Z0-9]$"
    return re.match(pattern, group_id) is not None
