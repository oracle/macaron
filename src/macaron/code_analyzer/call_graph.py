# Copyright (c) 2022 - 2024, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""This module contains classes to generate build call graphs for the target repository."""

from collections import deque
from collections.abc import Iterable
from typing import Any, Generic, TypeVar

Node = TypeVar("Node", bound="BaseNode")
# The documentation below for `TypeVar` is commented out due to a breaking
# change in Sphinx version (^=6.1.0).
# Reported at: https://github.com/oracle/macaron/issues/58.
# """This binds type ``Node`` to ``BaseNode`` and any of its subclasses.

# Therefore, any node of type ``Node`` that is stored in the call graph
# container will be a subtype of ``BaseNode``.
# """


class BaseNode(Generic[Node]):
    """This is the generic class for call graph nodes."""

    def __init__(self, caller: Node | None = None, node_id: str | None = None) -> None:
        """Initialize instance.

        Parameter
        ---------
        caller: Node | None
            The caller node.
        node_id: str | None
            The unique identifier of a node in the callgraph.
        """
        self.callee: list[Node] = []
        self.caller: Node | None = caller
        # Each node can have a model that summarizes certain properties for static analysis.
        # By default this model is set to None.
        self.model: Any = None
        self.node_id = node_id

    def add_callee(self, node: Node) -> None:
        """Add a callee to the current node.

        Parameters
        ----------
        node : Node
            The callee node.
        """
        self.callee.append(node)

    def has_callee(self) -> bool:
        """Check if the current node has callees.

        Returns
        -------
        bool
            Return False if there are no callees, otherwise True.
        """
        return bool(self.callee)


class CallGraph(Generic[Node]):
    """This is the generic class for creating a call graph."""

    def __init__(self, root: Node, repo_path: str) -> None:
        """Initialize instance.

        Parameters
        ----------
        root : Node
            The root call graph node.
        repo_path : str
            The path to the repo.
        """
        self.root = root
        self.repo_path = repo_path

    def get_root(self) -> Node:
        """Get the root node in the call graph.

        Returns
        -------
        Node
            The root node.
        """
        return self.root

    def bfs(self) -> Iterable[Node]:
        """Traverse the call graph in breadth first search order.

        Yields
        ------
        Node
            The traversed nodes.
        """
        queue: deque[Node] = deque()
        queue.extend(self.root.callee)
        visited = []
        while queue:
            node = queue.popleft()
            if node not in visited:
                queue.extend(node.callee)
                visited.append(node)
                yield node
