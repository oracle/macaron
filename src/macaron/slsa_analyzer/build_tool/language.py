# Copyright (c) 2024 - 2024, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""This module contains abstractions for build languages."""

from enum import Enum
from typing import Protocol, runtime_checkable


class BuildLanguage(str, Enum):
    """The supported build languages."""

    JAVA = "java"
    PYTHON = "python"
    GO = "go"
    JAVASCRIPT = "javascript"
    DOCKER = "docker"


@runtime_checkable
class Language(Protocol):
    """Interface of a language."""

    @property
    def lang_name(self) -> str:
        """Get the name of the language."""

    @property
    def lang_versions(self) -> list[str] | None:
        """Get the possible versions of the language."""

    @property
    def lang_distributions(self) -> list[str] | None:
        """Get the possible distributions of the language."""

    @property
    def lang_url(self) -> str | None:
        """Get the URL that provides information about the language distributions and versions."""
