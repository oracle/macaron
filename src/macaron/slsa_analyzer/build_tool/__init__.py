# Copyright (c) 2022 - 2022, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""The build_tool package contains the supported build tools for Macaron."""

from .base_build_tool import BaseBuildTool
from .gradle import Gradle
from .maven import Maven

# The list of supported build tools. The order of the list determine the order
# in which each build tool is checked against the target repository.
BUILD_TOOLS: list[BaseBuildTool] = [Gradle(), Maven()]
