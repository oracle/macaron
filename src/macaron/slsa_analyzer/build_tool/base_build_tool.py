# Copyright (c) 2022 - 2024, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""This module contains the BaseBuildTool class to be inherited by other specific Build Tools."""

import glob
import itertools
import json
import logging
import os
from abc import ABC, abstractmethod
from collections.abc import Iterable
from dataclasses import dataclass
from pathlib import Path
from typing import TypedDict

from macaron.code_analyzer.call_graph import BaseNode
from macaron.dependency_analyzer import DependencyAnalyzer
from macaron.slsa_analyzer.build_tool.language import BuildLanguage
from macaron.slsa_analyzer.checks.check_result import Confidence, Evidence, EvidenceWeightMap

logger: logging.Logger = logging.getLogger(__name__)


class BuildToolCommand(TypedDict):
    """This class is an abstraction for build tool commands storing useful contextual data for analysis."""

    #: The parsed build tool command. This command can be any bash command whose program name is the build tool.
    command: list[str]

    #: The name of the language to build the artifact.
    language: str

    #: The list of possible language version numbers.
    language_versions: list[str] | None

    #: The list of possible language distributions.
    language_distributions: list[str] | None

    #: The URL providing information about the language distributions and versions.
    language_url: str | None

    #: The relative path to the root CI file that ultimately triggers the command.
    ci_path: str

    #: The CI step object that calls the command.
    step_node: BaseNode

    #: The list of name of reachable variables that contain secrets."""
    reachable_secrets: list[str]

    #: The name of CI events that trigger the workflow running the build command.
    events: list[str] | None


def file_exists(path: str, file_name: str) -> bool:
    """Return True if a file exists in a directory.

    This method searches in the directory recursively.

    Parameters
    ----------
    path : str
        The path to search for the file.
    file_name : str
        The name of the file to search.

    Returns
    -------
    bool
        True if file_name exists else False.
    """
    pattern = os.path.join(path, "**", file_name)
    files_detected = glob.iglob(pattern, recursive=True)
    try:
        next(files_detected)
        return True
    except StopIteration:
        return False


@dataclass
class RuntimeOptions:
    """The class for build tool runtime configurations read from `defaults.ini`.

    Note that Macaron uses the options in this class to "run" a build tool.
    """

    #: The timeout used for running the build tool commands.
    build_timeout: float = 600


class BaseBuildTool(ABC):
    """This abstract class is used to implement Build Tools."""

    def __init__(self, name: str, language: BuildLanguage) -> None:
        """Initialize instance.

        Parameters
        ----------
        name : str
            The name of this build tool.
        language: BuildLanguage
            The name of the language used by the programs built by the build tool.
        """
        self.name = name
        self.language = language
        self.entry_conf: list[str] = []
        self.build_configs: list[str] = []
        self.package_lock: list[str] = []
        self.builder: list[str] = []
        self.packager: list[str] = []
        self.publisher: list[str] = []
        self.interpreter: list[str] = []
        self.interpreter_flag: list[str] = []
        self.build_arg: list[str] = []
        self.deploy_arg: list[str] = []
        self.ci_build_kws: dict[str, list[str]] = {
            "github_actions": [],
            "travis_ci": [],
            "circle_ci": [],
            "gitlab_ci": [],
            "jenkins": [],
        }
        self.ci_deploy_kws: dict[str, list[str]] = {
            "github_actions": [],
            "travis_ci": [],
            "circle_ci": [],
            "gitlab_ci": [],
            "jenkins": [],
        }
        self.build_log: list[str] = []
        self.wrapper_files: list[str] = []
        self.runtime_options = RuntimeOptions()

    def __str__(self) -> str:
        return self.name

    @abstractmethod
    def is_detected(self, repo_path: str) -> bool:
        """Return True if this build tool is used in the target repo.

        Parameters
        ----------
        repo_path : str
            The path to the target repo.

        Returns
        -------
        bool
            True if this build tool is detected, else False.
        """

    @abstractmethod
    def prepare_config_files(self, wrapper_path: str, build_dir: str) -> bool:
        """Prepare the necessary wrapper files for running the build.

        This method will return False if there is any errors happened during operation.

        Parameters
        ----------
        wrapper_path : str
            The path where all necessary wrapper files are located.
        build_dir : str
            The path of the build dir. This is where all files are copied to.

        Returns
        -------
        bool
            True if succeed else False.
        """

    @abstractmethod
    def load_defaults(self) -> None:
        """Load the default values from defaults.ini."""

    @abstractmethod
    def get_dep_analyzer(self, repo_path: str) -> DependencyAnalyzer:
        """Create a DependencyAnalyzer for the build tool.

        Parameters
        ----------
        repo_path: str
            The path to the target repo.

        Returns
        -------
        DependencyAnalyzer
            The DependencyAnalyzer object.
        """

    def get_build_dirs(self, repo_path: str) -> Iterable[Path]:
        """Find directories in the repository that have their own build scripts.

        This is especially important for applications that consist of multiple services.

        Parameters
        ----------
        repo_path: str
            The path to the target repo.

        Yields
        ------
        Path
            The relative paths from the repo path that contain build scripts.
        """
        config_paths: set[str] = set()
        for build_cfg in self.build_configs:
            config_paths.update(
                path
                for path in glob.glob(os.path.join(repo_path, "**", build_cfg), recursive=True)
                if self.is_detected(str(Path(path).parent))
            )

        list_iter = iter(sorted(config_paths, key=lambda x: (str(Path(x).parent), len(Path(x).parts))))
        try:
            cfg_path = next(list_iter)
            yield Path(cfg_path).parent.relative_to(repo_path)
            while next_item := next(list_iter):
                if next_item.startswith(str(Path(cfg_path).parent)):
                    continue
                cfg_path = next_item
                yield Path(next_item).parent.relative_to(repo_path)

        except StopIteration:
            pass

    def serialize_to_json(self, cmd: list[str]) -> str:
        """Convert a list of values to a json-encoded string so that it is easily parsable by later consumers.

        Parameters
        ----------
        cmd: list[str]
            List of command-line arguments.

        Returns
        -------
        str
            The list of command-line arguments as a json-encoded string.
        """
        return json.dumps(cmd)

    def is_build_command(self, cmd: list[str]) -> bool:
        """
        Determine if the command is a build tool command.

        Parameters
        ----------
        cmd: list[str]
            List of command-line arguments.

        Returns
        -------
        bool
            True if the command is a build tool command.
        """
        # Check for empty or invalid commands.
        if not cmd or not cmd[0]:
            return False
        # The first argument in a bash command is the program name.
        # So first check that the program name is a supported build tool name.
        # We need to handle cases where the first argument is a path to the program.
        cmd_program_name = os.path.basename(cmd[0])
        if not cmd_program_name:
            logger.debug("Found invalid program name %s.", cmd[0])
            return False

        build_tools = set(itertools.chain(self.builder, self.packager, self.publisher, self.interpreter))
        if any(tool for tool in build_tools if tool == cmd_program_name):
            return True

        return False

    def match_cmd_args(self, cmd: list[str], tools: list[str], args: list[str]) -> bool:
        """
        Check if the build command matches any of the tools and the command-line arguments.

        If build command's first element, which is the program name matches any of the `tools` names and any of its arguments
        match any of the arguments in `args`, this function returns True.

        Parameters
        ----------
        cmd: list[str]
            The command-line arguments.
        tools: list[str]
            The name of tools that will be matched with the program name in the bash command.
        args: list[str]
            The lit of arguments that should match with the bash command.

        Returns
        -------
        bool
            True if the provided command matches the tool and arguments.
        """
        cmd_program_name = os.path.basename(cmd[0])

        if cmd_program_name in tools:
            # Check the arguments in the bash command.
            # If there are no args expected for this build tool, accept the command.
            if not args:
                logger.debug("No build arguments required. Accept the %s command.", self.serialize_to_json(cmd))
                return True

            for word in cmd[1:]:
                # TODO: allow plugin versions in arguments, e.g., maven-plugin:1.6.8:deploy.
                if word in args:
                    logger.debug("Found the command %s.", self.serialize_to_json(cmd))
                    return True

        return False

    def infer_confidence_deploy_command(self, cmd: BuildToolCommand) -> Confidence:
        """
        Infer the confidence level for the deploy command.

        Parameters
        ----------
        cmd: BuildToolCommand
            The build tool command object.

        Returns
        -------
        Confidence
            The confidence level for the deploy command.
        """
        # Apply heuristics and assign weights and scores for the discovered evidence.
        # TODO: infer the scores based on existing data using probabilistic inference.
        # Initialize the map.
        evidence_weight_map = EvidenceWeightMap(
            [
                Evidence(name="reachable_secrets", found=False, weight=1),
                Evidence(name="ci_workflow_name", found=False, weight=2),
                Evidence(name="release_event", found=False, weight=2),
            ]
        )

        # Check if secrets are present in the caller job.
        if cmd["reachable_secrets"]:
            evidence_weight_map.update_result(name="reachable_secrets", found=True)

        # Check workflow names.
        deploy_keywords = ["release", "deploy", "publish"]
        for kw in deploy_keywords:
            if kw in os.path.basename(cmd["ci_path"]).lower():
                evidence_weight_map.update_result(name="ci_workflow_name", found=True)
                break

        # Check if the CI workflow is triggered by a release event.
        if cmd["events"] and "release" in cmd["events"]:
            evidence_weight_map.update_result(name="release_event", found=True)

        return Confidence.normalize(evidence_weight_map=evidence_weight_map)

    def is_deploy_command(
        self, cmd: BuildToolCommand, excluded_configs: list[str] | None = None
    ) -> tuple[bool, Confidence]:
        """
        Determine if the command is a deploy command.

        A deploy command usually performs multiple tasks, such as compilation, packaging, and publishing the artifact.
        This function filters the build tool commands that are called from the configuration files provided as input.

        Parameters
        ----------
        cmd: BuildToolCommand
            The build tool command object.
        excluded_configs: list[str] | None
            Build tool commands that are called from these configuration files are excluded.

        Returns
        -------
        tuple[bool, Confidence]
            Return True along with the inferred confidence level if the command is a deploy tool command.
        """
        # Check the language.
        if cmd["language"] is not self.language:
            return False, Confidence.HIGH
        # Some projects use a publisher tool and some use the build tool with deploy arguments.
        deploy_tool = self.publisher if self.publisher else self.builder

        if not self.match_cmd_args(cmd=cmd["command"], tools=deploy_tool, args=self.deploy_arg):
            return False, Confidence.HIGH

        # Check if the CI workflow is a configuration for a known tool.
        if excluded_configs and os.path.basename(cmd["ci_path"]) in excluded_configs:
            return False, Confidence.HIGH

        return True, self.infer_confidence_deploy_command(cmd=cmd)

    def is_package_command(
        self, cmd: BuildToolCommand, excluded_configs: list[str] | None = None
    ) -> tuple[bool, Confidence]:
        """
        Determine if the command is a packaging command.

        A packaging command usually performs multiple tasks, such as compilation and creating the artifact.
        This function filters the build tool commands that are called from the configuration files provided as input.

        Parameters
        ----------
        cmd: BuildToolCommand
            The build tool command object.
        excluded_configs: list[str] | None
            Build tool commands that are called from these configuration files are excluded.

        Returns
        -------
        tuple[bool, Confidence]
            Return True along with the inferred confidence level if the command is a build tool command.
        """
        # Check the language.
        if cmd["language"] is not self.language:
            return False, Confidence.HIGH

        builder = self.packager if self.packager else self.builder

        if not self.match_cmd_args(cmd=cmd["command"], tools=builder, args=self.build_arg):
            return False, Confidence.HIGH

        # Check if the CI workflow is a configuration for a known tool.
        if excluded_configs and os.path.basename(cmd["ci_path"]) in excluded_configs:
            return False, Confidence.HIGH

        return True, Confidence.HIGH
