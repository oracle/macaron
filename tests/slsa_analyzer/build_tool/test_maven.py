# Copyright (c) 2023 - 2024, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""This module tests the Maven build functions."""

from pathlib import Path

import pytest

from macaron.code_analyzer.call_graph import BaseNode
from macaron.slsa_analyzer.build_tool.base_build_tool import BuildToolCommand
from macaron.slsa_analyzer.build_tool.language import BuildLanguage
from macaron.slsa_analyzer.build_tool.maven import Maven
from tests.slsa_analyzer.mock_git_utils import prepare_repo_for_testing


@pytest.mark.parametrize(
    "mock_repo",
    [
        Path(__file__).parent.joinpath("mock_repos", "maven_repos", "has_parent_pom"),
        Path(__file__).parent.joinpath("mock_repos", "maven_repos", "no_parent_pom"),
        Path(__file__).parent.joinpath("mock_repos", "maven_repos", "no_pom"),
    ],
)
def test_get_build_dirs(snapshot: list, maven_tool: Maven, mock_repo: Path) -> None:
    """Test discovering build directories."""
    assert list(maven_tool.get_build_dirs(str(mock_repo))) == snapshot


@pytest.mark.parametrize(
    ("mock_repo", "expected_value"),
    [
        (Path(__file__).parent.joinpath("mock_repos", "maven_repos", "has_parent_pom"), True),
        (Path(__file__).parent.joinpath("mock_repos", "maven_repos", "no_parent_pom"), True),
        (Path(__file__).parent.joinpath("mock_repos", "maven_repos", "no_pom"), False),
    ],
)
def test_maven_build_tool(maven_tool: Maven, macaron_path: str, mock_repo: str, expected_value: bool) -> None:
    """Test the Maven build tool."""
    base_dir = Path(__file__).parent
    ctx = prepare_repo_for_testing(mock_repo, macaron_path, base_dir)
    assert maven_tool.is_detected(ctx.component.repository.fs_path) == expected_value


@pytest.mark.parametrize(
    (
        "command",
        "language",
        "language_versions",
        "language_distributions",
        "ci_path",
        "reachable_secrets",
        "events",
        "excluded_configs",
        "expected_result",
    ),
    [
        (
            ["mvn", "deploy"],
            BuildLanguage.JAVA,
            ["11", "17"],
            ["temurin"],
            ".github/workflows/release.yaml",
            [{"key", "pass"}],
            ["push"],
            ["codeql-analysis.yaml"],
            True,
        ),
        (
            ["mvn", "package"],
            BuildLanguage.JAVA,
            ["11", "17"],
            ["temurin"],
            ".github/workflows/release.yaml",
            [{"key", "pass"}],
            ["push"],
            ["codeql-analysis.yaml"],
            False,
        ),
        (
            ["mvn", "deploy"],
            BuildLanguage.PYTHON,
            ["11", "17"],
            ["temurin"],
            ".github/workflows/release.yaml",
            [{"key", "pass"}],
            ["push"],
            ["codeql-analysis.yaml"],
            False,
        ),
        (
            ["mvn", "deploy"],
            BuildLanguage.JAVA,
            ["11", "17"],
            None,
            ".github/workflows/release.yaml",
            [{"key", "pass"}],
            ["push"],
            ["codeql-analysis.yaml"],
            True,
        ),
        (
            ["mvn", "deploy"],
            BuildLanguage.JAVA,
            ["11", "17"],
            ["temurin"],
            ".github/workflows/maven.yaml",
            [{"key", "pass"}],
            ["push"],
            ["codeql-analysis.yaml"],
            True,
        ),
        (
            ["mvn", "deploy"],
            BuildLanguage.JAVA,
            ["11", "17"],
            ["temurin"],
            ".github/workflows/release.yaml",
            [],
            ["push"],
            ["codeql-analysis.yaml"],
            True,
        ),
        (
            ["mvn", "deploy"],
            BuildLanguage.JAVA,
            ["11", "17"],
            ["temurin"],
            ".github/workflows/release.yaml",
            [{"key", "pass"}],
            ["release"],
            ["codeql-analysis.yaml"],
            True,
        ),
        (
            ["mvn", "deploy"],
            BuildLanguage.JAVA,
            ["11", "17"],
            ["temurin"],
            ".github/workflows/maven.yaml",
            [{"key", "pass"}],
            ["push"],
            ["maven.yaml"],
            False,
        ),
        (
            ["npm", "publish"],
            BuildLanguage.JAVA,
            ["11", "17"],
            ["temurin"],
            ".github/workflows/maven.yaml",
            [{"key", "pass"}],
            ["push"],
            ["maven.yaml"],
            False,
        ),
    ],
)
def test_is_maven_deploy_command(
    maven_tool: Maven,
    command: list[str],
    language: str,
    language_versions: list[str],
    language_distributions: list[str],
    ci_path: str,
    reachable_secrets: list[str],
    events: list[str],
    excluded_configs: list[str] | None,
    expected_result: bool,
) -> None:
    """Test the deploy commend detection function."""
    result, _ = maven_tool.is_deploy_command(
        BuildToolCommand(
            command=command,
            language=language,
            language_versions=language_versions,
            language_distributions=language_distributions,
            language_url=None,
            ci_path=ci_path,
            step_node=BaseNode(),
            reachable_secrets=reachable_secrets,
            events=events,
        ),
        excluded_configs=excluded_configs,
    )
    assert result == expected_result


@pytest.mark.parametrize(
    (
        "command",
        "language",
        "language_versions",
        "language_distributions",
        "ci_path",
        "reachable_secrets",
        "events",
        "excluded_configs",
        "expected_result",
    ),
    [
        (
            ["mvn", "package"],
            BuildLanguage.JAVA,
            ["11", "17"],
            ["temurin"],
            ".github/workflows/release.yaml",
            [{"key", "pass"}],
            ["push"],
            ["codeql-analysis.yaml"],
            True,
        ),
        (
            ["mvn", "test"],
            BuildLanguage.JAVA,
            ["11", "17"],
            ["temurin"],
            ".github/workflows/release.yaml",
            [{"key", "pass"}],
            ["push"],
            ["codeql-analysis.yaml"],
            False,
        ),
        (
            ["mvn", "package"],
            BuildLanguage.PYTHON,
            ["11", "17"],
            ["temurin"],
            ".github/workflows/release.yaml",
            [{"key", "pass"}],
            ["push"],
            ["codeql-analysis.yaml"],
            False,
        ),
        (
            ["mvn", "package"],
            BuildLanguage.JAVA,
            ["11", "17"],
            None,
            ".github/workflows/release.yaml",
            [{"key", "pass"}],
            ["push"],
            ["codeql-analysis.yaml"],
            True,
        ),
        (
            ["mvn", "package"],
            BuildLanguage.JAVA,
            ["11", "17"],
            ["temurin"],
            ".github/workflows/maven.yaml",
            [{"key", "pass"}],
            ["push"],
            ["codeql-analysis.yaml"],
            True,
        ),
        (
            ["mvn", "package"],
            BuildLanguage.JAVA,
            ["11", "17"],
            ["temurin"],
            ".github/workflows/release.yaml",
            [],
            ["push"],
            ["codeql-analysis.yaml"],
            True,
        ),
        (
            ["mvn", "package"],
            BuildLanguage.JAVA,
            ["11", "17"],
            ["temurin"],
            ".github/workflows/release.yaml",
            [{"key", "pass"}],
            ["release"],
            ["codeql-analysis.yaml"],
            True,
        ),
        (
            ["mvn", "package"],
            BuildLanguage.JAVA,
            ["11", "17"],
            ["temurin"],
            ".github/workflows/maven.yaml",
            [{"key", "pass"}],
            ["push"],
            ["maven.yaml"],
            False,
        ),
        (
            ["npm", "publish"],
            BuildLanguage.JAVA,
            ["11", "17"],
            ["temurin"],
            ".github/workflows/maven.yaml",
            [{"key", "pass"}],
            ["push"],
            ["maven.yaml"],
            False,
        ),
        (
            ["mvn", "deploy"],
            BuildLanguage.JAVA,
            ["11", "17"],
            ["temurin"],
            ".github/workflows/maven.yaml",
            [{"key", "pass"}],
            ["push"],
            ["maven.yaml"],
            False,
        ),
    ],
)
def test_is_maven_package_command(
    maven_tool: Maven,
    command: list[str],
    language: str,
    language_versions: list[str],
    language_distributions: list[str],
    ci_path: str,
    reachable_secrets: list[str],
    events: list[str],
    excluded_configs: list[str] | None,
    expected_result: bool,
) -> None:
    """Test the packaging command detection function."""
    result, _ = maven_tool.is_package_command(
        BuildToolCommand(
            command=command,
            language=language,
            language_versions=language_versions,
            language_distributions=language_distributions,
            language_url=None,
            ci_path=ci_path,
            step_node=BaseNode(),
            reachable_secrets=reachable_secrets,
            events=events,
        ),
        excluded_configs=excluded_configs,
    )
    assert result == expected_result
