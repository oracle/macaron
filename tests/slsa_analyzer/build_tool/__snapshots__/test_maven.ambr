# serializer version: 1
# name: test_get_build_dirs[mock_repo0]
  list([
    PosixPath('.'),
  ])
# ---
# name: test_get_build_dirs[mock_repo1]
  list([
    PosixPath('sub_module_1'),
    PosixPath('sub_module_2'),
  ])
# ---
# name: test_get_build_dirs[mock_repo2]
  list([
  ])
# ---
